Add PowerShell command to setup Windows environment

mokagio opened the pull request:
---

- [ ] I have considered if this change warrants release notes and have added them to the appropriate section in the `CHANGELOG.md` if necessary.

mokagio committed:
Add script to set up Windows host for Node development

mokagio committed:
Ignore PowerShell scripts in test for executable property

Even if one runs `chmod u+x` on the file, the test fails.
Anyways, we don't care about PowerShell being executable from a Unix
point of view, because they should be used only under Windows.

mokagio committed:
Address RuboCop violation

I usually use Vim for Ruby files, where I have RuboCop configured.
Today, I used VS Code, where I _should have_ configured RuboCop but
evidently have not.

mokagio committed:
Update ShellCheck command to exclude PowerShell files

AliSoftware commented on Makefile:
```diff
@@ -12,7 +12,7 @@ buildkite-plugin-lint:
 shellcheck:
 	@echo ~~~ 🕵️ ShellCheck
-	$(docker_run) koalaman/shellcheck hooks/** bin/** --exclude=SC1071
+	$(docker_run) koalaman/shellcheck $(shell find hooks bin -type f -not -name "*.ps1") --exclude=SC1071
```
👍 

(I wonder if there's a `shellcheck` equivalent for PowerShell btw… though probably only a thing to look at when we'll start tacking #101 more officially/broadly?)

AliSoftware replied:
> I wonder if there's a shellcheck equivalent for PowerShell btw…

Found [this](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/issues/101#issuecomment-2189089950)… in case we want to look into it at some point

AliSoftware commented on tests/test-that-all-files-are-executable.rb:
```diff
@@ -11,10 +11,14 @@
-  Dir.children('bin').map { |f| File.new(File.join('bin', f)) }.each do |file|
-    it file.path do
-      expect(file.stat.executable?).to be true
+context 'All Unix Commands Should Be Executable' do
+  Dir
+    .children('bin')
+    # Ignore Windows PowerShell scripts
+    .reject { |f| f.end_with?('.ps1') }
```
👍

mokagio committed:
Log more information in `hooks/environment`

Context: I tried calling `add_ssh_key_to_agent` straight from the
Windows environment but it didn't work.

mokagio committed:
Print CI header upon running `add_ssh_key_to_agent`

mokagio committed:
Make `hooks/environment` properly set `PATH` in Windows

mokagio committed:
Remove OS-specific handling of paths in `hooks/environment`

mokagio committed:
Add debug `PATH` print statement in Windows PS1

mokagio committed:
Fix typo

mokagio committed:
Add `$PATH` retaining workaround to `refreshenv` call

mokagio committed:
Try to set dev mode for Windows machines

mokagio committed:
Use `CurrentUser` instead of `LocalMachine`

`LocalMachine` might not be allowed at a access right level.

See how this build failed
https://buildkite.com/automattic/beeper-desktop/builds/2897#0191979b-a067-4780-8a45-f051515841f1

mokagio committed:
Try different approach for developer mode

mokagio committed:
Debug print user roles in Windows

mokagio committed:
Implement conditional behavior for tar in Windows

mokagio committed:
Remove --no-symlinks

Doesn't work! ChatGPT what are you suggesting?!

https://buildkite.com/automattic/beeper-desktop/builds/2901#019197e0-da5d-47ec-84ec-8327b4aa5c7c/273-279

mokagio committed:
Restore using the same `tar` across platforms in `restore_cache`

Seems like all the flags did not make any difference in terms of whether
the symlink creation was attempted.

E.g. This build failed with the same symlink error as before, despite
having the flags
https://buildkite.com/automattic/beeper-desktop/builds/2907#0191982e-d1c5-402f-99be-2299321ccbe6

mokagio committed:
Run `nvm` setup at the end of the prepare Win host script

mokagio committed:
If now `.nvmrc` found, skip Node set up

mokagio committed:
Fix early exit check for `.nvmrc` in Windows

mokagio committed:
Do not open roles group

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 500 lin...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

mokagio committed:
Make `prepare_windows_host_for_node.ps1` executable

mokagio committed:
Merge remote-tracking branch 'origin/trunk' into mokagio/windows-utils

mokagio commented:
After months of this sitting here approved but with the tests failing, I finally got down to fixing it.

The reason for the test failure, i.e. for the new `ps1` not being added to the `$PATH` when running in CI (but working on my local, despite both using Docker) was that... the `ps1` was not marked as executable. It took ChatGPT to make me realize this. Thanks Open AI, I guess 😅 

See how https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/100/commits/647299c9a37735b8787cbc8ed5df54f11f51d700 fixes the `make test` CI failure.

I did a bit of force pushing to rebase this branch on `trunk` instead of https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/96 so it can be merged on `trunk` directly.

Given the only effective change since @AliSoftware approved this was changing the executable status of the `ps1` script, I'm going to cash in the approval and merge this so I can ship a new version.

The context for getting onto this change is that we'll need to set up Windows CI on the new Day One Electron app soon. I'm ready to jump on it, but I'd prefer if someone else had a go, to reduce the bus factor. With that in mind, it seemed appropriate to at least help and remove the oddity of having to use a custom branch in the CI setup.

mokagio commented:
I restored the branch so builds referencing it won't fail while we transition them to a stable version.

AliSoftware commented:
@mokagio you might also want to [add a CHANGELOG entry](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/100#issuecomment-2629702226) for this PR in a follow-up on before doing the next tag/release of this plugin
